File: Stack/Monotone Stack/1504. Count Submatrices With All Ones/solution.py
# O(M*M*N)
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        m, n = len(mat), len(mat[0])

        res = 0
        for i in range(m):
            height = [0] * n
            for j in range(i, m):
                for k in range(n):
                    if mat[j][k] == 0:
                        height[k] = 0
                    else:
                        height[k] += 1

                # 計算0到n這段有多少高度為h的合法矩形
                h = j-i+1
                cnt = width = 0
                for k in range(n):
                    if height[k] != h:
                        width = 0
                    else:
                        width += 1
                    cnt += width
                res += cnt
        return res
    
# O(MN)
class Solution:
    def numSubmat(self, mat: List[List[int]]) -> int:
        m, n = len(mat), len(mat[0])

        def histogram(arr):
            n = len(arr)
            counter = [0]*n
            stack = []

            for i in range(n):
                while stack and arr[stack[-1]] >= arr[i]:
                    stack.pop()

                if stack:
                    idx = stack[-1]
                    counter[i] = counter[idx]
                    counter[i] += arr[i] * (i-idx)
                else:
                    counter[i] = arr[i] * (i+1) # 0-indexed
                stack.append(i)
            return sum(counter)

        heights = [0] * n
        res = 0
        for i in range(m):
            for j in range(n):
                if mat[i][j] == 0:
                    heights[j] = 0
                else:
                    heights[j] += 1

            res += histogram(heights)
        return res

File: Stack/Monotone Stack/1504. Count Submatrices With All Ones/solution.cpp
class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m = mat.size(), n = mat[0].size();

        int res = 0;
        for (int top=0; top<m; top++) {
            vector<int> height(n, 0);
            for (int bot=top; bot<m; bot++) {
                for (int col=0; col<n; col++) {
                    if (mat[bot][col] == 1) {
                        height[col]++;
                    }
                }

                int h = bot-top+1;
                int width = 0;
                int cnt = 0;
                for (int col=0; col<n; col++) {
                    if (height[col] != h) {
                        width = 0;
                    } else {
                        width++;
                    }
                    cnt += width;
                }
                res += cnt;
            }
        }
        return res;
    }
};

class Solution {
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m=mat.size(),n=mat[0].size();

        for(int i=0;i<m;i++){
            for(int j=n-2;j>=0;j--){
                if(mat[i][j]==1){
                    mat[i][j]+=mat[i][j+1];
                }
            }
        }

        int res=0;
        for(int top=0;top<m;top++){
            for(int j=0;j<n;j++){
                int width=mat[top][j];
                for(int bot=top;bot<m;bot++){
                    width=min(width,mat[bot][j]);
                    res+=width;
                }
            }
        }

        return res;
    }
};

class Solution {
private:
    int histogram(vector<int>& height) {
        int n = height.size();
        vector<int> count(n, 0);

        stack<int> st;
        for (int i=0; i<n; i++) {
            while (!st.empty() && height[st.top()] >= height[i]) {
                st.pop();
            }

            if (!st.empty()) {
                int j = st.top();
                int w = i - j;
                count[i] = count[j] + height[i] * w;
            } else {
                int w = i+1;
                count[i] = height[i] * w;
            }
            st.push(i);
        }
    
        return accumulate(count.begin(), count.end(), 0);
    }
public:
    int numSubmat(vector<vector<int>>& mat) {
        int m=mat.size(),n=mat[0].size();

        vector<int> height(n, 0);
        int res = 0;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (mat[i][j] == 1) {
                    height[j]++;
                } else {
                    height[j] = 0;
                }
            }

            res += histogram(height);
        }
        return res;
    }
};